From 543f8470a9136a029551303b690c001ebeae4e2b Mon Sep 17 00:00:00 2001
From: Ding Wei <leo.ding@rock-chips.com>
Date: Wed, 17 Jun 2020 15:34:28 +0800
Subject: [PATCH] video: rockchip: mpp: fix issue iommu_release twice

arm_iommu_release_mapping will auto trigger when driver unregister,
so it should not manual release.

dump stack:
[   31.493866 ] ------------[ cut here  ]------------
[   31.493889 ] WARNING: CPU: 0 PID: 605 at lib/refcount.c:187
refcount_sub_and_test_checked+0x9c/0xb4
[   31.493898 ] refcount_t: underflow; use-after-free.
[   31.493907 ] Modules linked in: rk_vcodec(-)
[   31.493924 ] CPU: 0 PID: 605 Comm: rmmod Not tainted 4.19.111 #2047
[   31.493932 ] Hardware name: Generic DT based system
[   31.493957 ] [<b010f408>] (unwind_backtrace) from [<b010b96c>]
(show_stack+0x10/0x14)
[   31.493975 ] [<b010b96c>] (show_stack) from [<b07ed264>]
(dump_stack+0x90/0xa4)
[   31.493992 ] [<b07ed264>] (dump_stack) from [<b0125f64>]
(__warn+0xfc/0x114)
[   31.494006 ] [<b0125f64>] (__warn) from [<b0125fc0>]
(warn_slowpath_fmt+0x44/0x68)
[   31.494020 ] [<b0125fc0>] (warn_slowpath_fmt) from [<b0398f40>]
(refcount_sub_and_test_checked+0x9c/0xb4)
[   31.494036 ] [<b0398f40>] (refcount_sub_and_test_checked) from
[<b0116730>] (arm_iommu_release_mapping.part.5+0x20/0x48)
[   31.494050 ] [<b0116730>] (arm_iommu_release_mapping.part.5) from
[<b0117408>] (arch_teardown_dma_ops+0x2c/0x38)
[   31.494065 ] [<b0117408>] (arch_teardown_dma_ops) from [<b047a87c>]
(device_release_driver_internal+0xf4/0x19c)
[   31.494081 ] [<b047a87c>] (device_release_driver_internal) from
[<b047a980>] (driver_detach+0x44/0x80)
[   31.494094 ] [<b047a980>] (driver_detach) from [<b04796b4>]
(bus_remove_driver+0x4c/0xa0)
[   31.494120 ] [<b04796b4>] (bus_remove_driver) from [<af005ff8>]
(mpp_service_exit+0x3c/0x1044 [rk_vcodec])
[   31.494140 ] [<af005ff8>] (mpp_service_exit [rk_vcodec]) from
[<b019c97c>] (sys_delete_module+0x190/0x1f0)
[   31.494154 ] [<b019c97c>] (sys_delete_module) from [<b0101000>]
(ret_fast_syscall+0x0/0x4c)
[   31.494162 ] Exception stack(0xed62ffa8 to 0xed62fff0)
[   31.494172 ] ffa0:                   00025a04 765f6b72 ae982c08
00000880 00000000 ae982e88
[   31.494183 ] ffc0: 00025a04 765f6b72 65646f63 00000081 000c46a4
00000000 a6f49000 00000000
[   31.494191 ] ffe0: ae982c00 ae982bf0 000258f8 a6ea31b0
[   31.494288 ] ---[ end trace 6bdf082e7c4aa12d  ]---

Change-Id: Ib4a057087b1521a5591c12909b6d78b5f90f3882
Signed-off-by: Ding Wei <leo.ding@rock-chips.com>
---
 drivers/video/rockchip/mpp/mpp_iommu.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/drivers/video/rockchip/mpp/mpp_iommu.c b/drivers/video/rockchip/mpp/mpp_iommu.c
index 3354549da83e..f9b8d3d2267d 100644
--- a/drivers/video/rockchip/mpp/mpp_iommu.c
+++ b/drivers/video/rockchip/mpp/mpp_iommu.c
@@ -428,12 +428,6 @@ mpp_iommu_probe(struct device *dev)
 
 int mpp_iommu_remove(struct mpp_iommu_info *info)
 {
-#ifdef CONFIG_ARM_DMA_USE_IOMMU
-	struct dma_iommu_mapping *mapping;
-
-	mapping = to_dma_iommu_mapping(info->dev);
-	arm_iommu_release_mapping(mapping);
-#endif
 	iommu_group_put(info->group);
 	platform_device_put(info->pdev);
 
-- 
2.35.3

